Create CSS Modules migration post (#5159)
sourcegraph/about: blogposts/2022/css-modules-migration.md

## Migration tracking challenges

The Frontend Platform team started the migration by manually converting global styles to CSS modules for a single, recently developed feature. We quickly noticed that despite some initial progress, we had some questions that slowed us down:

1. How can we know whether CSS Modules are continuously used for new features added to the project?
2. What’s our current progress? What percentage of global styles have already been converted to CSS Modules?
3. Given the current rate of changes, how much time will it take to migrate all remaining styles to CSS Modules?

We knew we could search the codebase manually for relevant files and make conclusions based on that, or we could write a script to do it for us. But an ideal solution would give a clear, at-a-glance indication of where we were at the moment, and be easy for us to maintain. Thankfully, another product team at Sourcegraph had developed the exact solution to our problem!

“Code Insights reveals high-level information about your codebase, based on both how your code changes over time and its current state.” – [source](https://docs.sourcegraph.com/code_insights).

> Code Insights entered Beta in August 2021, and we happily started using it to track the migration progress. As of today, [Code Insights is now Generally Available](LINK TO ANNOUNCEMENT).

Author: TODO: Add link

Author: ```suggestion
Code Insights entered Beta in August 2021, and we happily started using it to track the migration progress. As of today, [Code Insights is now Generally Available](/blog/announcing-code-insights/).
```

<figure>
  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/code-insights-ga-blogs/migration-to-css-modules-insight.png" alt="Abstract syntax tree">
  <figcaption>Count of CSS files that are global vs module files.</figcaption>

> </figure>

Author: TODO: Add link to live insight

Author: @umpox @valerybugakov Do you have a link I can use here? We wanted to embed it but that's not possible on the blog yet.

(Also, just noticed the alt text is wrong, I will fix that)

Reviewer: A link directly to the code insight? We don't unfortunately, this insight isn't publicly accessible and AFAIK we can't directly link the code insights right now

Reviewer 2: Confirmed with Code Insights team. It's not possible to reference or embed this insight.

## Migration execution challenges

Global migrations are challenging for platform teams. After spending time to migrate another feature to CSS Modules, we discovered some execution obstacles in our way:

1. The migration work is quite repetitive, and we had to do a ton of it because the size of the codebase is pretty significant: ~400 files to migrate.
2. We wanted to make the migration as smooth as possible for product teams. Asking them to allocate a substantial amount of their time to work on the migration was an option but not ideal.
3. It's not very exciting to work on for an extended period of time, but dragging out the migration would mean leaving the codebase inconsistent for too long. New engineers might be confused about which approach to use, even if we document it.

## Codemods to the rescue

> A codemod is an automated change to source code, which helps platform teams execute global refactorings faster and with higher confidence. Because codemods are usually written in the same language as the project, they can understand its subtleties and complexities. It makes them perfect for executing large-scale migration where find-and-replace functionality is insufficient. A codemod generally includes the following steps:

Author: @valerybugakov We talked about briefly explaining the difference between codemods and Batch Changes but I don't see that anywhere. Can you help add it please?

Reviewer 2: Hey @rebeccadee, since then, I've updated this paragraph with a detailed description of what codemods do. We do not reference Batch Changes in the blog post, and this name is specific to the Sourcegraph product; I feel it's redundant to raise this topic here. What do you think?

Author: @valerybugakov Agreed, thank you!

2. 🥷 Having a way to migrate 95% of use cases automatically greatly reduces the time commitment needed and allows the platform team to complete the refactoring behind the scenes without asking product teams for help.
3. 🏎 Faster migration means that the codebase stays inconsistent for much less time.

## Codemod implementation

To programmatically migrate global styles to CSS modules, codemod script does three major things:

### 1. Find pairs of files to transform.

Finding a React component file with the corresponding global-styles file is easy in the Sourcegraph codebase because React components have colocated styles with a predictable filename. E.g., styles for `Button.tsx` are defined in `Button.scss` in the same folder.

### 2. Convert global styles file into a CSS module.

It’s a multistep operation:

- First, get rid of the BEM notation used for CSS selectors and remove redundant nesting required to ensure selector uniqueness.

```css
.insights-dashboard {
  flex: none;

  &__wrapper {
    display: flex;
  }
}
```

⬇

```css
.insights-dashboard {
  flex: none;
}

.wrapper {
  display: flex;
}

> ```

Reviewer: LGTM!

- To achieve this, we used [PostCSS](https://github.com/postcss/postcss)—a ​tool for transforming styles with JS plugins. The resulting script is similar to [the postcss-nested plugin](https://github.com/postcss/postcss-nested), with some adjustments to remove the BEM notation.
- Save a list of CSS Module classes in memory for later reference in the React component. It is done by using utility functions provided by [the css-modules-loader-core package](https://github.com/css-modules/css-modules-loader-core). They give us a list of classes available in a CSS Module that we can map ourselves to global CSS classes that we just transformed. For our toy example, it would look like this:

> ```css
> const classMapping = {
>   'insights-dashboard':'insights-dashboard','insights-dashboard__wrapper': 'wrapper';
> }
> ```

Reviewer: ```suggestion
```js
const classMapping = {
  'insights-dashboard': 'insights-dashboard',
  'insights-dashboard__wrapper': 'wrapper',
}
```
```

> ```

Author: Same here, is this right?
<img width="869" alt="image7" src="https://user-images.githubusercontent.com/82806852/155334143-26dc97d5-aee9-46a3-ac5f-1c764d5b4be2.png">

- As a final touch, add `.module.scss` to the filename required for build-tools to interpret this file as a CSS Module.

### 3. Replace references to global classes in the React component using a list of CSS Modules classes preserved from the previous step.

To manipulate Typescript AST, we used [ts-morph](https://github.com/dsherret/ts-morph) —TypeScript Compiler API wrapper for static analysis and programmatic code changes. Relying on this package API, the codemod script iterates over all string literals in the React component AST and searches for global classes processed in the previous step.

Here’s the AST generated for our small example. Explore it yourself using [astexplorer](https://astexplorer.net/#/gist/eca630c9f5464e6b027ec13aac91711a/4295e325b8cdad474f9d4f559248f534e709356e).

<figure>
  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/code-insights-ga-blogs/insights-dashboard-ast.png" alt="Insights dashboard AST"
</figure>

Codemod replaces every string literal match with reference to the corresponding CSS Module class.

> ```css
> <div className="insights-dashboard__wrapper" />
> ```

Reviewer: ```suggestion
```jsx
<div className="insights-dashboard__wrapper" />
```
```

⬇

> ```css
> <div className={styles.wrapper} />
> ```

Reviewer: ```jsx
<div className={styles.wrapper} />
```
```suggestion
```jsx
<div className={styles.wrapper} />
```
```

> ```

Author: Is this correct?
![image4](https://user-images.githubusercontent.com/82806852/155334282-0595b8b5-5c68-4782-a137-5d80f72c84e5.png)

At this point, the transformation is complete, and the script outputs information about CSS classes not used in the React component. It allowed us to remove dead code in the migration process.

The codemod script can be used as a CLI tool by supplying a list of files to transform:

> `$ yarn transform --write --transform ./globalCssModule.ts ./sourcegraph/**/*.tsx`

Reviewer: ```suggestion
```sh
$ yarn transform --write --transform ./globalCssModule.ts ./sourcegraph/**/*.tsx
```
```

> `$ yarn transform --write --transform ./globalCssModule.ts ./sourcegraph/**/*.tsx`

Author: Was trying to capture this, need to check if correct
![image1](https://user-images.githubusercontent.com/82806852/155334361-dfdef497-e99a-44fb-8f90-ae6353b179e7.png)

Author: ```suggestion
`$ yarn transform --write --transform ./globalCssToCssModule.ts ./sourcegraph/**/*.tsx`
```

Author: Applied in batched suggestions

<figure>
  <img src="https://storage.googleapis.com/sourcegraph-assets/blog/code-insights-ga-blogs/migration-progress.svg" alt="Intermediate migration progress"
</figure>

## Collaboration with contractors

Another challenge we faced was that we kept breaking our own rules: In quite a few places, we found we were using styles scoped outside of the respective component. Handling those cases required manual intervention from the engineers after applying the codemod.

To keep executing the migration behind the scenes, we started working with contractors to help us handle cases that the codemod could not address. It was another productivity lever akin to the codemod that immensely helped us to focus on critical problems while keeping the migration process going in the background.

> <blockquote class="twitter-tweet" tw-align-center"><p lang="en" dir="ltr">Nothing like using <a href="https://twitter.com/sourcegraph?ref_src=twsrc%5Etfw">@sourcegraph</a> to build <a href="https://twitter.com/sourcegraph?ref_src=twsrc%5Etfw">@sourcegraph</a>—we&#39;re migrating from global CSS to CSS Modules and our frontend platform team is using Code Insights to track migration progress: <a href="https://t.co/1lRqYjLiwz">pic.twitter.com/1lRqYjLiwz</a></p>&mdash; Beyang Liu (@beyang) <a href="https://twitter.com/beyang/status/1454937232368832522?ref_src=twsrc%5Etfw">October 31, 2021</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script>

Author: TODO: Fix alignment

Reviewer 2: Fixed

---
title: 'How we migrated entirely to CSS Modules using codemods and Sourcegraph Code Insights'
author: Valery Bugakov
description: 'How our Frontend Platform team used codemods to automate a challenging global migration to CSS modules, and Code Insights to track and communicate progress.'

> publishDate: 2022-02-24T00:00-07:00

Author: ```suggestion
publishDate: 2022-03-10T00:00-07:00
```

2. Global CSS rules depend on the order of the code, and it's difficult to ensure that the order is preserved when implementing global changes like the application redesign. Moving files around could result in a slightly different import order that affects a tiny bit of UI. Debugging such small changes can significantly slow teams down and makes DX a nightmare.
3. There was no robust and configured way to identify unused CSS code not linked to any HTML elements. In the process of the redesign, we occasionally uncovered such styles. These styles added no value to our codebase or our product.

To overcome these issues, we decided that, before diving into the redesign, we should find a way to embrace scoped styles instead of adding new styles to the global scope. At the same time, we needed to ensure that this solution could be integrated with the current approach effectively. The definition of success was a combination of the following outcomes:

1. Modular CSS is enforced for new design system components.
2. It works well with existing global styles and UI elements.
3. We have a clear migration path that doesn’t interfere with feature development.

## CSS Modules solution

> We researched popular solutions available in the open source universe: trade-offs between CSS in JS, regular CSS, and hybrid solutions. Some offered more powerful features, but we decided to use a solution that we knew we could easily adopt — CSS modules.

Author: ```suggestion
We researched popular solutions available in the open source universe: trade-offs between CSS in JS, regular CSS, and hybrid solutions. Some offered more powerful features, but we decided to use a solution that we knew we could easily adopt—CSS modules.
```

### What are CSS Modules?

CSS modules are great because they help you write reusable components with isolated styles. According to the [repo](https://github.com/css-modules/css-modules), CSS modules are:

“CSS files in which all class names and animation names are scoped locally by default.”

> With CSS modules, CSS classes should be referenced in JavaScript file via explicit binding to the styles file:

Author: ```suggestion
With CSS modules, CSS classes should be referenced in the JavaScript file via explicit binding to the styles file:
```
This might not be quite right but we're definitely missing a word here! @valerybugakov
